core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,8 +62,14 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
```

Comment:
Heh, yeah can you see the long collapsed conversation above? that was indeed the concern. Here, you would have to have source whose entire class was not thread-safe (i.e. you can't concurrently access two unrelated instances), which would be pretty broken. My contention is that such a source would break in other places in Spark, so trying to fix it here doesn't do anything. It also sounds like there's no known instance of this (?) so that's why I'm not sure if this is a meaningful option.